searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_blueprint.cpp

```diff
@@ -73,20 +74,31 @@ NearestNeighborBlueprint::NearestNeighborBlueprint(const queryeval::FieldSpec& f
         _dist_fun = nns_index->distance_function();
     }
     uint32_t est_hits = _attr_tensor.getNumDocs();
-    if (_approximate && nns_index) {
-        est_hits = std::min(target_num_hits, est_hits);
-    }
     setEstimate(HitEstimate(est_hits, false));
     set_want_global_filter(true);
 }
 
 NearestNeighborBlueprint::~NearestNeighborBlueprint() = default;
 
 void
-NearestNeighborBlueprint::set_global_filter(std::shared_ptr<BitVector> global_filter)
+NearestNeighborBlueprint::set_global_filter(const GlobalFilter &global_filter)
 {
-    // XXX do something with global_filter here
-    (void) global_filter;
+    _global_filter = global_filter.shared_from_this();
+    auto nns_index = _attr_tensor.nearest_neighbor_index();
+    if (_approximate && nns_index) {
+        uint32_t est_hits = _attr_tensor.getNumDocs();
+        if (_global_filter->has_filter()) {
+            uint32_t max_hits = _global_filter->filter()->countTrueBits();
+            if (max_hits * 10 < est_hits) {
```

Comment:
This threshold should probably be configurable, but that we can do in a later PR.